[nnc] Update bounds overlap analysis to handles partial constant bounds

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74658
* #74346

This diff updates the bounds overlap analysis to handle this case: when the diff between the bounds is a constant, only one of them needs to be constant and positive for the bounds to not overlap. 

Differential Revision: [D35102669](https://our.internmc.facebook.com/intern/diff/D35102669)

navahgar committed:
[nnc] Update bounds overlap analysis to identify non-overlaps even with symbolic bounds

[ghstack-poisoned]

navahgar committed:
Update on "[nnc] Update bounds overlap analysis to identify non-overlaps even with symbolic bounds"


This diff updates the bounds overlap analysis to handle two cases:
1. When the diff between the bounds is a constant, only one of them needs to be constant and positive for the bounds to not overlap. 
2. When the diff between the bounds has a variable, we perform a conservative approximation to figure out cases that are guaranteed to be positive. See comments in file `bounds_overlap.cpp` for a detailed explanation of this approximation.

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35102669).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -6,6 +6,41 @@ namespace jit {
+// The expression could involve multiple subtract terms and negative
+// immediate values and end up being positive. But even if one of them is
+// present we conservatively assume that it can be negative.
+bool isPositiveExpr(ExprPtr e) {
```
Naming nit: `isPositive` sounds like the function is accurate with both `true` and `false`, while in fact we only guarantee the `true`. I.e. `!isPositive` can not be used as `isNegative`. Maybe  `mustBePositive`, `alwaysPositive` or something along these lines?

navahgar replied:
Sounds good. Changed to `mustBePositive`.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -14,16 +49,43 @@ OverlapKind boundOverlap(Bound a, Bound b) {
+  if (lowDiff->isConstant()) {
     int low = immediateAs<int>(lowDiff);
+    // No overlap.
+    if (low > 0) {
```
This condition only works when the "gap" is more than 0, while the condition below seem to be true for 0. If that's so, can we make it consistent? Also, do we even need this if? Can it all be handled by a single `isPositiveExpr` check?

navahgar replied:
Good point. I didn't think about the difference earlier. Thanks for pointing that out.

We need to ensure that it is positive because 0 means there is some overlap. I have modified the code to look for 0 as well now and identify only the guaranteed positive cases. PTAL.

ZolotukhinM replied:
When we have `a.start -- a.end` do we actually include `a.end`? Not allowing 0 seems to result in a hairy code now, so maybe we should just make our ranges to not include the last value.

navahgar replied:
Unfortunately, our bounds data structure is set up to include `a.end`. I'm afraid changing that would involve touching a lot of the mem analysis code base. 

In the current setup, we should not be allowing 0. Here is the approximation that I currently have:
 
The given expression is guaranteed to be positive if all these conditions hold:
   * does not include any "subtract" term
   * does not include any "div" term
   * all immediate values are positive (> 0)

I see that this is not quite enough since an expression can get to 0 with mod, shift, xor, etc. 

I'm thinking that we should be even more conservative and allow the expressions that only contain:
  * "add" and "mul" terms
  * all immediate values are positive.
Though, implementing this is going to involve checking all the cases that should not be allowed, which is frustrating.

What do you think?

navahgar replied:
I wasn't able to convince myself that this handles all the corner cases. Also, implementing this (checking for all such possible cases that could lead to a zero or negative value) is cumbersome and errorprone. So, I'm removing the checks with variables in this.

ZolotukhinM commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -1969,10 +1969,7 @@ TEST_F(Kernel, FuseLoopsWithVariableConcatDim) {
-  # CHECK-NEXT: for (int64_t k
-  # CHECK: for (int64_t j
-  # CHECK-NEXT: for (int64_t k
+  # CHECK-NOT: for (int64_t j
```
We started to fuse this case, right? Can we also now add a test for a case which we shouldn't be fusing?

navahgar replied:
Sure, added another test for that.

ZolotukhinM reviewed (approved):
Overall looks good, a couple of comments inline.

navahgar committed:
Update on "[nnc] Update bounds overlap analysis to identify non-overlaps even with symbolic bounds"


This diff updates the bounds overlap analysis to handle two cases:
1. When the diff between the bounds is a constant, only one of them needs to be constant and positive for the bounds to not overlap. 
2. When the diff between the bounds has a variable, we perform a conservative approximation to figure out cases that are guaranteed to be positive. See comments in file `bounds_overlap.cpp` for a detailed explanation of this approximation.

Differential Revision: [D35102669](https://our.internmc.facebook.com/intern/diff/D35102669)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35102669).

navahgar commented:
@ZolotukhinM PTAL again since I modified the overlap criteria to exclude 0 cases.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -6,6 +6,73 @@ namespace jit {
+//
+// The only way for the given expression to be zero is if it includes at least
+// one of:
+//     * "subtract" term.
+//     * "div" term.
+//     * a zero immediate value.
```
What if we have `100 + (-100)`?

navahgar replied:
The comment was not corresponding to the actual code. The actual code was checking both the negative and zero cases together. So, we will conclude that `100 + (-100)` is indeed overlapping.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -6,6 +6,73 @@ namespace jit {
+  AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TYPE_CASE);
+#undef TYPE_CASE
+  {
+    auto imms = NodeFinder<BoolImm>::find(e);
```
Just curious, why do we care about `BoolImm`s in arithmetic expressions?

navahgar replied:
Only because a bool being 0 can make the overall expression 0.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -6,6 +6,73 @@ namespace jit {
+//
+// The only way for the given expression to be negative is if it includes
+// at least one of:
+//     * "subtract" term.
```
In theory bitwise ops plus cast can also result in a negative value. But I only mention it for the sake of completeness, I'm not sure we really need to add them here. Same with shifts for zero values.

ZolotukhinM reviewed (approved):
👍

navahgar committed:
Update on "[nnc] Update bounds overlap analysis to identify non-overlaps even with symbolic bounds"


This diff updates the bounds overlap analysis to handle two cases:
1. When the diff between the bounds is a constant, only one of them needs to be constant and positive for the bounds to not overlap. 
2. When the diff between the bounds has a variable, we perform a conservative approximation to figure out cases that are guaranteed to be positive. See comments in file `bounds_overlap.cpp` for a detailed explanation of this approximation.

Differential Revision: [D35102669](https://our.internmc.facebook.com/intern/diff/D35102669)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35102669).

navahgar committed:
Update on "[nnc] Update bounds overlap analysis to handles partial constant bounds"


This diff updates the bounds overlap analysis to handle this case: when the diff between the bounds is a constant, only one of them needs to be constant and positive for the bounds to not overlap. 

Differential Revision: [D35102669](https://our.internmc.facebook.com/intern/diff/D35102669)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35102669).

navahgar committed:
Update on "[nnc] Update bounds overlap analysis to handles partial constant bounds"


This diff updates the bounds overlap analysis to handle this case: when the diff between the bounds is a constant, only one of them needs to be constant and positive for the bounds to not overlap. 

Differential Revision: [D35102669](https://our.internmc.facebook.com/intern/diff/D35102669)

[ghstack-poisoned]
